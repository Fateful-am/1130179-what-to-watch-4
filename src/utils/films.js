export const MovieInfoMode = {
  CARD: `card`,
  PROMO: `promo`,
  OVERVIEW: `overview`,
  DETAIL: `detail`,
  REVIEWS: `reviews`
};

export const getMovies = (sourceMovies, sliceFrom, count) => {
  return sourceMovies.slice(sliceFrom, sliceFrom + count).map((movie) => {
    return {
      id: movie.id,
      title: movie.title,
      posterImage: movie.posterImage,
    };
  });
};

export const getMoviePromoMode = (sourceMovies, index) => {
  const movie = sourceMovies[index];
  return {
    id: movie.id,
    title: movie.title,
    genre: movie.genre,
    posterImage: movie.posterImage,
    backgroundImage: movie.backgroundImage,
    released: movie.released
  };
};
