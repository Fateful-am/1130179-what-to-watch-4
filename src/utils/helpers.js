import moment from 'moment';

export const convertToImageName = (movieName) => {
  return movieName.toLowerCase().split(` `).join(`-`).split(`:`).join(``);
};

export const getMovieById = (movies, id) => {
  const movieIndex = movies.findIndex((movie) => movie.id === id);
  if (movieIndex > -1) {
    return movies[movieIndex];
  }
  return null;
};

export const formatDuration = (duration) => {
  if (duration < 60) {
    return moment.utc(moment.duration(duration, `minutes`).asMilliseconds()).format(`m[m]`);
  }
  return moment.utc(moment.duration(duration, `minutes`).asMilliseconds()).format(`H[h] m[m]`);
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};


export const smallMovieCards = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`,
  `Aviator`,
  `We need to talk about Kevin`,
  `What We Do in the Shadows`,
  `Revenant`,
  `Johnny English`,
  `Shutter Island`,
  `Pulp Fiction`,
  `No Country for Old Men`,
  `Snatch`,
  `Moonrise Kingdom`,
  `Seven Years in Tibet`,
  `Midnight Special`,
  `War of the Worlds`,
  `Dardjeeling Limited`,
  `Orlando`,
  `Mindhunter`,
  `Midnight Special`
];
