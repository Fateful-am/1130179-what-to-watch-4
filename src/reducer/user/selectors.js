import NameSpace from "../name-space";
import {wtwLocalStorage} from '../../utils/helpers';

const NAME_SPACE = NameSpace.USER;


export const getAuthorizationStatus = (state) => {
  const localStorageAuthStatus = wtwLocalStorage.getAuthStatus();
  return localStorageAuthStatus
    ? localStorageAuthStatus
    : state[NAME_SPACE].authorizationStatus;
};

export const getLoginErrorMessage = (state) => {
  return state[NAME_SPACE].loginErrorMessage;
};

export const getUserData = (state) => {
  return state[NAME_SPACE].userData;
};
