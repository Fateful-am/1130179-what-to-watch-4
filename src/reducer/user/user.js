import {extend} from '../../utils/helpers';
import {ActionCreator as MovieActionCreator} from '../movie/movie';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  userData: {},
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_USERDATA: `SET_USERDATA`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },

  setUserData: (userData) => {
    return {
      type: ActionType.SET_USERDATA,
      payload: userData,
    };
  }
};

const convertToLocalUserData = (serverUserData) => {
  if (serverUserData) {
    return {
      id: serverUserData[`id`],
      email: serverUserData[`email`],
      name: serverUserData[`name`],
      avatarUrl: serverUserData[`avatar_url`],
    };
  }
  return {
    userData: {}
  };
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_USERDATA:
      return extend(state, {
        userData: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.setUserData(convertToLocalUserData(response.data)));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.email,
      password: authData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.setUserData(convertToLocalUserData(response.data)));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(MovieActionCreator.gotoMain());
      });
  },
};

export {
  AuthorizationStatus,
  ActionCreator,
  ActionType,
  Operation,
  reducer,
};
