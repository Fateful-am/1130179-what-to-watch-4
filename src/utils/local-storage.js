import {extend} from './helpers';

const ItemsType = {
  AUTH_STATUS: `authStatus`,
  LAST_URL: `lastUrl`,
};

export default class LocalStorage {
  constructor(key, storage) {
    this._storage = storage;
    this._storeKey = key;
  }

  _getItems() {
    try {
      return JSON.parse(this._storage.getItem(this._storeKey)) || {};
    } catch (err) {
      return {};
    }
  }

  _getItem(item) {
    const store = this._getItems();
    return store[item];
  }

  _setItem(item, value) {
    const store = this._getItems();
    this._storage.setItem(
        this._storeKey,
        JSON.stringify(extend(store, {
          [item]: value,
        }))
    );
  }

  getAuthStatus() {
    return this._getItem(ItemsType.AUTH_STATUS);
  }

  setAuthStatus(authStatus) {
    this._setItem(ItemsType.AUTH_STATUS, authStatus);
  }

  getLastUrl() {
    return this._getItem(ItemsType.LAST_URL) || `/`;
  }

  setLastUrl(url) {
    this._setItem(ItemsType.LAST_URL, url);
  }
}
