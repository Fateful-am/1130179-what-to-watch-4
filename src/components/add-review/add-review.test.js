import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import AddReview from './add-review';
import {TEST_DATA} from '../../utils/test-data';
import NameSpace from '../../reducer/name-space';
import {extend} from '../../utils/helpers';
import {AuthorizationStatus} from '../../reducer/user/user';

const mockStore = configureStore([]);

it(`Render AddReview Page`, () => {
  const store = mockStore({
    [NameSpace.MOVIE]: TEST_DATA.initialStoreMovieState,
    [NameSpace.USER]: extend(TEST_DATA.initialStoreUserState, {
      authorizationStatus: AuthorizationStatus.AUTH,
    })
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <AddReview
            movie={TEST_DATA.promoMovie}
            onBreadcrumbsBackClick={()=>{}}
          >
          </AddReview>
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        }
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
