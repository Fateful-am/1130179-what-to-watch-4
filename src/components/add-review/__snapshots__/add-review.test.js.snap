// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render AddReview Page 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="img/bg-the-grand-budapest-hotel.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/8"
              onClick={[Function]}
            >
              The Grand Budapest Hotel
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
          onClick={[Function]}
        >
          <img
            alt="User avatar"
            height="63"
            src="https://4.react.pages.academyundefined"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="The Grand Budapest Hotel poster"
        height="327"
        src="img/the-grand-budapest-hotel-poster.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  />
</section>
`;
